test/test_custom_ops.py

```diff
@@ -678,6 +680,134 @@ def foo(x: Tensor, y: int) -> Tensor:
+        def _(x, y):
+            return x - y
+
+        z = f(x, y)
+        self.assertEqual(z, x - y)
+
+        with self.assertRaises(RuntimeError):
+            with torch._subclasses.fake_tensor.FakeTensorMode():
+                x = torch.randn(3)
+                f(x, y)
+
+        @f.impl_abstract
+        def _(x, y):
+            return torch.empty_like(x)
+
+        with torch._subclasses.fake_tensor.FakeTensorMode():
+            x = torch.randn(3)
+            z = f(x, y)
+            self.assertEqual(z.shape, x.shape)
+
+    @skipIfTorchDynamo("recursive dynamo")
+    @unittest.skipIf(IS_WINDOWS, "torch.compile doesn't work on windows")
+    @unittest.skipIf(
+        sys.version_info >= (3, 12), "torch.compile is not supported on python 3.12+"
+    )
+    def test_blackbox_compile(self):
+        @torch.library.opaque_op(mutated_args=())
+        def custom_linear(x: Tensor, weight: Tensor, bias: Tensor) -> Tensor:
+            return (x @ weight.t()) + bias
+
+        @custom_linear.impl_abstract
+        def _(x, weight, bias):
+            assert x.dim() == 2
+            assert weight.dim() == 2
+            assert bias.dim() == 1
+            assert x.shape[1] == weight.shape[1]
+            assert weight.shape[0] == bias.shape[0]
+            assert x.device == weight.device
+            return x.new_empty(x.size(0), weight.size(0))
+
+        x = torch.randn(2, 2)
+        weight = torch.randn(2, 2)
+        bias = torch.randn(2)
+        out = torch.compile(custom_linear, backend="eager", fullgraph=True)(
+            x, weight, bias
+        )
+        assert torch.allclose(out, torch.nn.functional.linear(x, weight, bias))
+
+    def test_blackbox_replaces(self):
+        @opaque_op(mutated_args=())
+        def f(x: Tensor) -> Tensor:
+            return x.sin()
+
+        x = torch.randn(3)
+        y = f(x)
+        self.assertEqual(y, x.sin())
+
+        @opaque_op(mutated_args=())
+        def f(x: Tensor) -> Tensor:
+            return x.cos()
+
+        y = f(x)
+        self.assertEqual(y, x.cos())
+
+    @unittest.skipIf(not TEST_CUDA, "requires CUDA")
+    def test_blackbox_split_device(self):
+        @opaque_op(mutated_args=(), types="cpu")
+        def f(x: Tensor) -> Tensor:
+            return x.sin()
+
+        @f.impl("cuda")
+        def _(x: Tensor) -> Tensor:
+            return x.cos()
+
+        x = torch.randn(3)
+        y = f(x)
+        self.assertEqual(y, x.sin())
+        x = x.cuda()
+        y = f(x)
+        self.assertEqual(y, x.cos())
+
+    @unittest.skipIf(not TEST_CUDA, "requires CUDA")
+    def test_blackbox_multi_types(self):
+        @opaque_op(mutated_args=(), types=("cpu", "cuda"))
+        def f(x: Tensor) -> Tensor:
+            return x.sin()
+
+        x = torch.randn(3)
+        y = f(x)
+        self.assertEqual(y, x.sin())
+        x = x.cuda()
+        y = f(x)
+        self.assertEqual(y, x.sin())
+
+    def test_blackbox_disallows_output_aliasing(self):
```

Comment:
Huh, so is the test done at runtime? Seems like it could potentially be a source of overhead

Reply:
Yes, the test is done at runtime. We could add it to opcheck instead to avoid overhead.

Reply:
Btw, autograd.Function does similar checks at runtime, so, I'm not worried about performance right now. If it becomes a problem we can move the check to opcheck instead of runtime.